go/vt/sqlparser/ast.go

```diff
@@ -4845,6 +4846,109 @@ func (node *Order) walkSubtree(visit Visit) error {
 	)
 }
 
+type FrameUnit int
+
+const (
+	// RangeUnit is the mode of specifying frame in terms of logical range (e.g. 100 units cheaper).
+	RangeUnit FrameUnit = iota
+	// RowsUnit is the mode of specifying frame in terms of physical offsets (e.g. 1 row before etc).
+	RowsUnit
+)
+
+// Frame represents a window Frame clause
+type Frame struct {
+	Unit   FrameUnit
+	Extent *FrameExtent
+}
+
+// Format formats the node.
+func (node *Frame) Format(buf *TrackedBuffer) {
+	if node == nil {
+		return
+	}
+	switch node.Unit {
+	case RangeUnit:
+		buf.Myprintf(" RANGE")
+	case RowsUnit:
+		buf.Myprintf(" ROWS")
+	}
+}
+
+func (node *Frame) walkSubtree(visit Visit) error {
+	if node == nil {
+		return nil
+	}
+	return Walk(
+		visit,
+		node.Extent,
+	)
+}
+
+// FrameExtent defines the start and end bounds for a window frame
+type FrameExtent struct {
+	Start, End *FrameBound
+}
+
+// Format formats the node.
+func (node *FrameExtent) Format(buf *TrackedBuffer) {
+	if node == nil {
+		return
+	}
+	if node.End != nil {
+		buf.Myprintf(" BETWEEN")
+		node.Start.print(buf)
+		buf.Myprintf(" AND")
+		node.End.print(buf)
+	} else {
+		node.Start.print(buf)
+	}
+}
+
+func (node *FrameExtent) walkSubtree(visit Visit) error {
+	return nil
+}
+
+type BoundType int
+
+const (
+	// CurrentRow represents the current row position
+	CurrentRow BoundType = iota
+	// UnboundedFollowing includes all rows after CURRENT ROW in the active partition
+	UnboundedFollowing
+	// UnboundedPreceding includes all rows before CURRENT ROW in the active partition
+	UnboundedPreceding
+	// ExprPreceding matches N rows or value sets before the CURRENT ROW
+	ExprPreceding
+	// ExprFollowing matches N rows or value sets after the CURRENT ROW
+	ExprFollowing
+)
+
+// FrameBound defines one direction of row or range inclusion
+type FrameBound struct {
+	Expr Expr
+	Type BoundType
+}
+
+// Format formats the node.
+func (node *FrameBound) print(buf *TrackedBuffer) {
```

Comment:
should `FrameBound` implement `Node`? the Format function would be empty, but it's useful in the AST

Reply:
Why would the format function be empty? It has a textual representation, right?

Reply:
was confused, realize now that node printing is recursive